Src/WitsmlExplorer.Frontend/contexts/filter.tsx

```diff
@@ -1,22 +1,103 @@
+import NavigationType from "./navigationType";
 
-interface Filter {
-  wellName: string;
+export interface Filter {
+  name: string;
   isActive: boolean;
   objectGrowing: boolean;
   wellLimit: number;
 }
 
 export const EMPTY_FILTER: Filter = {
-  wellName: "",
+  name: "",
   isActive: false,
   objectGrowing: false,
   wellLimit: 30
 };
 
-const filterWellsOnWellName = (wells: Well[], wellNameFilter: string) => {
-  if (!wellNameFilter || wellNameFilter === "") return wells;
-  return wells.filter((well: Well) => well.name.toLowerCase().includes(wellNameFilter.toLowerCase()));
+interface FilterContextProps {
+  selectedFilter: Filter;
+  updateSelectedFilter: (partialFilter: Partial<Filter>) => void;
+}
+
+export const FilterContext = React.createContext<FilterContextProps>({} as FilterContextProps);
+
+export function FilterContextProvider({ children }: React.PropsWithChildren) {
+  const [selectedFilter, setSelectedFilter] = React.useState<Filter>(EMPTY_FILTER);
+
+  const updateSelectedFilter = React.useCallback((partialFilter: Partial<Filter>) => {
+    setSelectedFilter((prevFilter) => ({
+      ...prevFilter,
+      ...partialFilter
+    }));
+  }, []);
+
+  const contextValue: FilterContextProps = React.useMemo(
+    () => ({
+      selectedFilter,
+      updateSelectedFilter
+    }),
+    [selectedFilter, updateSelectedFilter]
+  );
+
+  return <FilterContext.Provider value={contextValue}>{children}</FilterContext.Provider>;
+}
+
+export interface FilterOptions {
+  matchOnlyWell?: boolean; // Only return wells where the well matches the filter
+  filterWellbores?: boolean; // Filter the wellbores of the wells that don't match. Only effective if matchOnlyWell is set to false.
+  dispatchNavigation?: (action: NavigationAction) => void; //A function to dispatch an action to expand tree nodes every time the filter changes.
+}
+
+export const DEFAULT_FILTER_OPTIONS: FilterOptions = {
+  matchOnlyWell: false,
+  filterWellbores: false
+};
+
+const filterOnWellOrWellboreName = (wells: Well[], wellNameFilter: string, filterOptions: FilterOptions) => {
+  const { matchOnlyWell, filterWellbores, dispatchNavigation } = filterOptions;
+  if (!wellNameFilter || wellNameFilter === "") {
+    if (filterOptions.dispatchNavigation) {
+      const expandTreeNodes: ExpandTreeNodesAction = { type: NavigationType.ExpandTreeNodes, payload: { nodeIds: [] } };
+      dispatchNavigation(expandTreeNodes);
+    }
+    return wells;
+  }
+
+  const regexPattern = wellNameFilter
+    .replace(/[.+^${}()|[\]\\]/g, "\\$&") // Escape unsupported regex-symbols (except for * and ?)
+    .replace(/\*/g, ".*") // Replace * with .* to match any characters
+    .replace(/\?/g, "."); // Replace ? with . to match any single character
+
+  const regex = new RegExp(`${regexPattern}`, "i");
+
+  const filteredWells: Well[] = [];
+  const treeNodesToExpand: string[] = [];
+
+  for (const well of wells) {
+    if (regex.test(well.name)) {
```

Comment:
Deleting a well that is selected and matches the current filter will cause a crash on this line
1. Create a well called "well123"
2. Search for "well123"
3. Select the well
4. Delete the well
![image](https://github.com/equinor/witsml-explorer/assets/66632214/f9d16695-695a-490a-87d2-6431a417d0ac)